dashboard/lib/pd/payment/payment_calculator_base.rb

```diff
@@ -0,0 +1,244 @@
+      end
+    end
+
+    private
+
+    # Calculates raw attendance totals per teacher based on a supplied set of session attendance summaries.
+    # @param session_attendance_summaries [Array<SessionAttendanceSummary>]
+    #   Summary of teacher attendance for each session
+    # @return [Hash{Integer => TeacherAttendanceTotal}]
+    #   Map of teacher id to raw attendance totals for that teacher.
+    def calculate_raw_teacher_attendance(session_attendance_summaries)
+      {}.tap do |raw_attendance_per_teacher|
+        session_attendance_summaries.each do |session_attendance_summary|
+          hours = session_attendance_summary.hours
+          session_attendance_summary.teacher_ids.each do |teacher_id|
+            if raw_attendance_per_teacher[teacher_id]
+              raw_attendance_per_teacher[teacher_id].add_session hours
+            else
+              raw_attendance_per_teacher[teacher_id] = TeacherAttendanceTotal.new 1, hours
+            end
+          end
+        end
+      end
+    end
+
+    # Constructs workshop payment from a workshop summary
+    # @param workshop_summary [WorkshopSummary]
+    # @return [WorkshopPayment, nil] Payment details, or nil if the workshop is not qualified for payment.
+    def construct_workshop_payment(workshop_summary)
+      return nil unless qualified? workshop_summary.total_teacher_attendance_days
+
+      WorkshopPayment.new.tap do |workshop_payment|
+        workshop_payment.summary = workshop_summary
+        workshop_payment.type = get_payment_type workshop_summary.workshop
+        workshop_payment.amounts = calculate_payment_amounts workshop_summary
+        workshop_payment.total = workshop_payment.amounts.values.reduce(0, :+)
+      end
+    end
+
+    # Constructs teacher summaries (including payment) for each teacher in the workshop.
+    # @param workshop_summary [WorkshopSummary]
+    # @param raw_teacher_attendance [Hash{Integer => TeacherAttendanceTotal}]
+    #   Map of teacher id to raw attendance totals for that teacher.
+    # @return [Array<TeacherSummary>] summary for each teacher.
+    def construct_teacher_summaries(workshop_summary, raw_teacher_attendance)
+      enrollments_by_teacher_id = {}
+      enrolled_teachers_by_id = {}
+      workshop_summary.workshop.enrollments.each do |enrollment|
+        teacher = enrollment.user
+        if teacher
+          enrollments_by_teacher_id[teacher.id] = enrollment
+          enrolled_teachers_by_id[teacher.id] = teacher
+        end
+      end
+
+      # Generate a teacher summary for all teachers in raw attendance.
+      raw_teacher_attendance.keys.map do |teacher_id|
+        enrollment = enrollments_by_teacher_id[teacher_id]
+        teacher = enrolled_teachers_by_id[teacher_id] || User.find(teacher_id)
+
+        TeacherSummary.new.tap do |teacher_summary|
+          raw_attendance = raw_teacher_attendance[teacher_id]
+
+          teacher_summary.workshop_summary = workshop_summary
```

Comment:
Good call. I wrote constructors for `WorkshopSummary`, `WorkshopPayment`, `TeacherSummary`, and `TeacherPayment` each with required named parameters and made those fields read-only.